### PR TITLE
Runtime type hashconsing

### LINKED ISSUE #12: Reimplement tuple and function glb functions iteratively

### PR DESCRIPTION
**Pull request by deyaaeldeen**
Runtime hash-consing of all types. This makes structure equality on types reduced to pointer equality. I am not 100% satisfied with the solution because it allocates types whether or not the allocation will be used. I will work on avoiding this allocation in another PR, but I believe this will mean I will have to move the implementation of `greatest_lower_bound` to the C land to combine it with hashconsing.

Also fixes #12 by making them iterative and inlining them.

### CODE REVIEW

File: src/language/syntax.rkt [hunk 1]
```diff
@@ -168,6 +168,7 @@
 
 
 (define-syntax-rule (op$ o a ...) (Op 'o (list a ...)))
+(define-syntax-rule (app$ cl a ...) (App-Code cl (list a ...)))
```
Comment (akuhlens) [inline src/language/syntax.rkt]: This should probably be named `appc$` or `app-code$` to distinguish it from a macro that creates the `App` form.

File: src/casts/specify-representation.rkt [hunk 2]
```diff
@@ -60,492 +70,601 @@ but a static single assignment is implicitly maintained.
...
+(: get-hashcons-types! (-> (Code-Label Uid)))
+(define (get-hashcons-types!)
+  (: make-code! (-> (Code-Label Uid)))
+  (define (make-code!)
+    (define-track-next-uid!$ hashcons-types)
+    (define hashcons-types-label (Code-Label hashcons-types))
+    (define err-msg (Quote "specify-representation/hashcons-types: switch failure"))
+    (define hashcons-types-c : D0-Code
+      (code$ (ty)
+        (If (op$ <= ty TYPE-MAX-ATOMIC-RT-VALUE)
```
Comment (akuhlens) [inline src/casts/specify-representation.rkt]: You could avoid right creep here with the cond$ macro.

File: src/language/syntax.rkt (hunk 1 above)
Comment (deyaaeldeen) [reply src/language/syntax.rkt]: Done.

File: src/casts/specify-representation.rkt (hunk 2 above)
Comment (deyaaeldeen) [reply src/casts/specify-representation.rkt]: It does not help much but done.

**Comment by deyaaeldeen** [comment]
I added two more commits. I fixed a bug when hashcodes overflow and added a test case to force the hashtable to resize at runtime. I constructed a benchmark that creates a `Tuple` type with 800 dyn nodes and casted the value node by node to the full static type. The runtime of the benchmark is the same for `master` and this PR. Increasing the number of nodes will cause a stack overflow in `master`. I believe the PR now is ready to be merged.

**Pull request merged**